src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -0,0 +1,396 @@
+         */
+        SKIPPED,
+
+        /**
+         * when the compaction was aborted for the CFS because the CF got dropped in the meantime
+         */
+        ABORTED,
+
+        /**
+         * compaction tasks were completed for the CFS
+         */
+        COMPLETED
+    }
+
+    private final DebuggableScheduledThreadPoolExecutor checkExecutor;
+
+    /**
+     * Each time we start a compaction, the CFS is added to this set; it is removed when the compaction finishes
+     * or when it is cancelled
+     */
+    private final Multiset<ColumnFamilyStore> ongoingCompactions = ConcurrentHashMultiset.create();
+
+    /**
+     * CFSs for which a compaction was requested mapped to the promise returned to the requesting code
+     */
+    private final ConcurrentMap<ColumnFamilyStore, List<SettableFuture<RequestResult>>> compactionRequests = new ConcurrentHashMap<>();
+
+    private final AtomicInteger currentlyBackgroundUpgrading = new AtomicInteger(0);
+
+    private final Random random = new Random();
+
+    private final DebuggableThreadPoolExecutor compactionExecutor;
+
+    private final ActiveOperations activeOperations;
+
+
+    BackgroundCompactionRunner(DebuggableThreadPoolExecutor compactionExecutor, ActiveOperations activeOperations)
+    {
+        this(compactionExecutor, new DebuggableScheduledThreadPoolExecutor("BackgroundTaskExecutor"), activeOperations);
+    }
+
+    @VisibleForTesting
+    BackgroundCompactionRunner(DebuggableThreadPoolExecutor compactionExecutor, DebuggableScheduledThreadPoolExecutor checkExecutor, ActiveOperations activeOperations)
+    {
+        this.compactionExecutor = compactionExecutor;
+        this.checkExecutor = checkExecutor;
+        this.activeOperations = activeOperations;
+    }
+
+    /**
+     * Marks a CFS for compaction. Since marked, it will become a possible candidate for compaction. The mark will be
+     * cleared when we actually run the compaction for CFS.
+     *
+     * @return a promise which will be completed when the mark is cleared
+     */
+    Future<RequestResult> requestCompaction(ColumnFamilyStore cfs)
+    {
+        logger.trace("Requested background compaction for {}.{}", cfs.getKeyspaceName(), cfs.getTableName());
+
+        if (checkExecutor.isShutdown())
+        {
+            logger.info("Executor has been shut down, background compactions check will not be scheduled");
+            return Futures.immediateCancelledFuture();
+        }
+
+        SettableFuture<RequestResult> futureResult = SettableFuture.create();
+        compactionRequests.compute(cfs, (ignored, current) -> {
+            if (current == null)
+                current = Collections.synchronizedList(new ArrayList<>());
+            current.add(futureResult);
+            return current;
+        });
+
+        if (!maybeScheduleNextCheck())
+        {
+            futureResult.cancel(false);
```

Comment:
Let's cancel all tasks in the set on shutdown instead.

Reply:
we would have to synchronize `requestCompaction` and `shutdown` in that case, right?